embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/EmbraceSpanImpl.kt

```diff
@@ -114,24 +119,37 @@ internal class EmbraceSpanImpl(
         }
     }
 
-    override fun addEvent(name: String, timestampMs: Long?, attributes: Map<String, String>?): Boolean {
-        if (eventCount.get() < MAX_EVENT_COUNT && inputsValid(name, attributes)) {
-            val newEvent = EmbraceSpanEvent.create(
+    override fun addEvent(name: String, timestampMs: Long?, attributes: Map<String, String>?): Boolean =
+        recordEvent(name = name, attributes = attributes) {
+            EmbraceSpanEvent.create(
                 name = name,
                 timestampMs = timestampMs?.normalizeTimestampAsMillis() ?: openTelemetryClock.now().nanosToMillis(),
                 attributes = attributes
             )
-            synchronized(eventCount) {
-                if (eventCount.get() < MAX_EVENT_COUNT && isRecording) {
-                    events.add(newEvent)
-                    eventCount.incrementAndGet()
-                    return true
-                }
-            }
         }
 
-        return false
-    }
+    override fun recordException(exception: Throwable, attributes: Map<String, String>?): Boolean =
+        recordEvent(name = EXCEPTION_EVENT_NAME, attributes = attributes) {
+            val eventAttributes = mutableMapOf<String, String>()
+            if (attributes != null) {
+                eventAttributes.putAll(attributes)
+            }
+            exception.javaClass.canonicalName?.let { type ->
+                eventAttributes[exceptionType.key] = type
+            }
+            exception.message?.let { message ->
+                eventAttributes[exceptionMessage.key] = message
+            }
+
+            eventAttributes[exceptionStacktrace.key] =
+                serializer.toJson(exception.stackTrace.map(StackTraceElement::toString).take(200).toList(), List::class.java)
```

Comment:
Would suggest `take()` should go first to avoid unnecessary object instantiation for humungous stacktraces